rfcs/008-map-indexes.md

```diff
@@ -0,0 +1,202 @@
+1. The emitted keys are encoded then added to the `?VIEW_MAP_RANGE` with the emitted keys and value stored
+1. The `?VIEW_ROW_COUNT` is incremented
+1. The `?VIEW_KV_SIZE` is increased
+1. If the document was deleted and was previously in the view, the previous keys for the document are read from `?VIEW_ID_RANGE` and then cleared from the `?VIEW_MAP_RANGE`. The Row count and size count are also decreased.
+1. If the document is being updated and was previously added to the index, then he previous keys for the document are read from `?VIEW_ID_RANGE` and then cleared from the `?VIEW_MAP_RANGE` and then the index is updated with the latest emitted keys and value.
+
+### Key ordering
+
+FoundationDB orders key by byte value which is not how CouchDB orders keys. To maintain CouchDB's view collation, a type value will need to be prepended to each key so that the correct sort order of null < boolean < numbers < strings < arrays < objects is maintained.
+
+In CouchDB 2.x, strings are compared via ICU. The way to do this with FoundationDB is that for every string an ICU sort string will be generated upfront and added to the key.
+
+### Emitting document
+
+In a map function, it is possible to emit the full document as the value, this will cause an issue if the document size is larger than FDB’s value limit of 100 KB. We can handle this in two possible ways. The first is to keep the hard limit of only allowing 100 KB value to be emitted, so if a document exceeds that CouchDB will return an error. This is the preferred option.
+
+The second option is to detect that a map function is emitting the full document and then add in a foreign key reference back to the document subspace. The issue here is that CouchDB would only be able to return the latest version of the document, which would cause consistency issues when combined with the `update=false` argument.
+
+A third option would be to split the document across multiple keys in FoundationDB. This will still be limited by the transaction size limit.
+
+### Index building
+
+An index will be built and updated via a [background job worker](https://github.com/apache/couchdb-documentation/blob/master/rfcs/007-background-jobs.md). When a request for a view is received, the request process will add a job item onto the background queue for the index to be updated. A worker will take the item off the queue and update the index. Once the index has been built, the background job server will notify the request that the index is up to date. The request process will then read from the index and return the results. This process can also be optimised in two ways. Firstly, using a new couch_events system to listen for document changes in a database and then adding indexing jobs to the queue to keep indexes warm. The second optimisation is if the index only requires a small update, rather update the index in the HTTP request process instead of doing the work via the background queue.
```

Comment:
I'm a bit dubious about adding indexing workloads to HTTP workers if the deployment decides to split out different workers for HTTP and indexing. Is there an alternative via creating some kind of "fast path" for couch_jobs where the HTTP worker is able to directly contact an index worker to get it to perform the work rather than work via FoundationDB?